### PR TITLE
[nrf fromtree] Lwm2m multi-instanced firmware update object

### PR DESCRIPTION
**Pull request by jarlamsa**
As https://github.com/zephyrproject-rtos/zephyr/pull/41402 didn't make it to the upmerge, this PR brings the commits to sdk-zephyr. 
Also https://github.com/zephyrproject-rtos/zephyr/pull/41598 is brought in this PR.

### CODE REVIEW

**Review by rlubos — APPROVED** [review]

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware.c [hunk 1]
```diff
@@ -66,41 +73,47 @@ static struct lwm2m_engine_obj_field fields[] = {
...
-extern int lwm2m_firmware_start_transfer(char *package_uri);
+extern int lwm2m_firmware_start_transfer(uint16_t obj_inst_id, char *package_uri);
 #endif
 
+uint8_t lwm2m_firmware_get_update_state_inst(uint16_t obj_inst_id)
+{
+	return update_state[obj_inst_id];
+}
+
 uint8_t lwm2m_firmware_get_update_state(void)
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:92]: This function should be removed as it's not needed anymore.

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware.c [hunk 2]
```diff
@@ -113,67 +126,80 @@ void lwm2m_firmware_set_update_state(uint8_t state)
...
+{
+	lwm2m_firmware_set_update_state_inst(0, state);
+}
+
+uint8_t lwm2m_firmware_get_update_result_inst(uint16_t obj_inst_id)
+{
+	return update_result[obj_inst_id];
 }
 
 uint8_t lwm2m_firmware_get_update_result(void)
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:149]: This function should be removed as it's not needed anymore.

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]
Those two functions are something I should have spotted myself earlier, sorry for that.

**Comment by rlubos** [comment]
Just note, we shouldn't modify commit cherry-picked from the upstream, as the changes would be lost anyway on the next upmerge. If there are some shortcomings in the code, we should aim to fix them in the upstream instead, and cherry-pick the fixes if urgent.

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware.c (hunk 1 above)
Comment (jarlamsa) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:92]: It is needed for the backwards compatibility

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware.c (hunk 2 above)
Comment (jarlamsa) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:149]: It is needed for the backwards compatibility

**Comment by jarlamsa** [comment]
Manifest PR: https://github.com/nrfconnect/sdk-nrf/pull/6665
@rlubos thanks for guidance.

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware.c (hunk 1 above)
Comment (VeijoPesonen) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:92]: Then it should be kept on the header file too.

**Review by VeijoPesonen — APPROVED** [review]

**Comment by jarlamsa** [comment]
The missing function declarations were merged to zephyr main, cherry-picked that commit. @rlubos is there something still needed for this and the manifest PR to get these merged?

**Comment by rlubos** [comment]
Let's just wait for the CI here, I see the manifest update is green already so I can pull it in

**Pull request merged**